crates/wasi-keyvalue/src/provider/inmemory.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{generated::wasi::keyvalue::store::KeyResponse, Error, Host};
+use async_trait::async_trait;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+#[derive(Default)]
+pub(crate) struct InMemory {
+    store: Arc<Mutex<HashMap<String, Vec<u8>>>>,
+}
+
+#[async_trait]
+impl Host for InMemory {
+    async fn get(&mut self, key: String) -> Result<Option<Vec<u8>>, Error> {
+        let store = self.store.lock().unwrap();
+        Ok(store.get(&key).cloned())
+    }
+
+    async fn set(&mut self, key: String, value: Vec<u8>) -> Result<(), Error> {
+        let mut store = self.store.lock().unwrap();
+        store.insert(key, value);
+        Ok(())
+    }
+
+    async fn delete(&mut self, key: String) -> Result<(), Error> {
+        let mut store = self.store.lock().unwrap();
+        store.remove(&key);
+        Ok(())
+    }
+
+    async fn exists(&mut self, key: String) -> Result<bool, Error> {
+        let store = self.store.lock().unwrap();
+        Ok(store.contains_key(&key))
+    }
+
+    async fn list_keys(&mut self, cursor: Option<u64>) -> Result<KeyResponse, Error> {
+        let store = self.store.lock().unwrap();
+        let keys: Vec<String> = store.keys().cloned().collect();
+        let cursor = cursor.unwrap_or(0) as usize;
+        let keys_slice = &keys[cursor..];
+        Ok(KeyResponse {
+            keys: keys_slice.to_vec(),
+            cursor: None,
+        })
+    }
+
+    async fn increment(&mut self, key: String, delta: u64) -> Result<u64, Error> {
+        let mut store = self.store.lock().unwrap();
+        let value = store.entry(key.clone()).or_insert(vec![0]);
+        let current_value = String::from_utf8(value.clone())
+            .unwrap_or_else(|_| "0".to_string())
+            .parse::<u64>()
+            .unwrap_or(0);
```

Comment:
It loosk like this is silently ignoring non-integer previous values and converting them to 0?

Reply:
Fixed, now, non-integer data conversion errors will be returned directly.